Display `BitField[Enum]` in docs to distinguish from `Enum`

dalexeev opened the pull request:
* ~~Treat `BitField<T>` as an `int` rather than an enum `T` in GDScript.~~ Fixed by #75691.
* Change display of bitfields in Editor Help as `BitField[T]`.

![](https://user-images.githubusercontent.com/47700418/223954931-6e6f78b2-e75b-4d35-944b-c1adce621bae.png)

I highlighted the `BitField` with a different color and added a hint, since `BitField` is not a type in GDScript.

![](https://user-images.githubusercontent.com/47700418/223955249-8e2d9022-915a-46cc-bd7d-4d1ffc668282.png)

![](https://user-images.githubusercontent.com/47700418/223955477-b8e95baf-cdf7-4ab7-87c2-fdf6401cd61d.png)

akien-mga commented:
Looks fairly good to me, though I'm a bit concerned about introducing the `BitField[Enum]` syntax which might be confusing, especially since it's not supported in GDScript.

For example, how would users know which type is actually returned by `button_mask` in GDScript? Is it `MouseButtonMask`, is it some `BitField[]` that would raise a syntax error if they try to use it as type hint, or is it a plain `int`?

dalexeev commented:
> For example, how would users know which type is actually returned by `button_mask` in GDScript? Is it `MouseButtonMask`, is it some `BitField[]` that would raise a syntax error if they try to use it as type hint, or is it a plain `int`?

Now it is plain `int`, not `SomeEnum`.

> I'm a bit concerned about introducing the `BitField[Enum]` syntax which might be confusing

Yes, I'm a little worried about this too, so I used a different color (same as `void`) and added a tooltip. A little confusion is unavoidable, but the current state is already confusing. `ENUM_VALUE_A | ENUM_VALUE_B` is `int` in GDScript. And the `SomeEnum` type hint raises a warning when an `int` is passed to it.

![](https://user-images.githubusercontent.com/47700418/223973234-51ba1a10-7313-44ee-b385-c178c983d341.png)

akien-mga commented on editor/editor_help.cpp:
```diff
@@ -245,6 +246,13 @@ void EditorHelp::_add_type(const String &p_type, const String &p_enum) {
 			class_desc->add_text("[");
+		} else if (is_bitfield) {
+			class_desc->push_color(Color(theme_cache.type_color, 0.5));
+			class_desc->push_hint(TTR("This value is a bitmask of the following flags."));
```
```suggestion
			class_desc->push_hint(TTR("This value is integer composed as a bitmask of the following flags."));
```
Not sure that's the best phrasing, but I think we should make it explicit it's an `int`.

akien-mga commented on editor/editor_help.cpp:
```diff
@@ -245,6 +246,13 @@ void EditorHelp::_add_type(const String &p_type, const String &p_enum) {
 			class_desc->add_text("[");
+		} else if (is_bitfield) {
+			class_desc->push_color(Color(theme_cache.type_color, 0.5));
+			class_desc->push_hint(TTR("This value is integer composed as a bitmask of the following flags."));
```
```suggestion
			class_desc->push_hint(TTR("This value is an integer composed as a bitmask of the following flags."));
```

dalexeev commented:
Online docs uses C++ style:

<details>
<summary>Images</summary>

![](https://user-images.githubusercontent.com/47700418/224104658-a6c57c97-61bd-4ec4-8cd0-490058127058.png)
![](https://user-images.githubusercontent.com/47700418/224105022-2fe95d2c-a537-486a-95f0-61cf6c6e673d.png)
![](https://user-images.githubusercontent.com/47700418/224105589-93d8934e-62e6-4587-b427-cd3def4c7e9c.png)
![](https://user-images.githubusercontent.com/47700418/224107686-34dba6b0-74f8-490f-8ca8-a746bb956ee9.png)

</details>

vnen commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -4557,6 +4557,9 @@ GDScriptParser::DataType GDScriptAnalyzer::type_from_property(const PropertyInfo
 						result.is_constant = false;
 					}
 				}
+			} else if ((p_property.usage & PROPERTY_USAGE_CLASS_IS_BITFIELD) && p_property.class_name != StringName()) {
+				// BitField isn't supported (yet?).
+				result.is_constant = false;
```
This `else` is not needed. The only reason it sets `result.is_constant = false` here is because `make_enum_type()` makes it `true`, but in the `else` branch it is not called.

Calinou reviewed (approved):
Code looks good to me. I think the current style isn't too bad – we may have to iterate on it later, but it needs testing in real world conditions first :slightly_smiling_face:

dalexeev commented:
Rebased. Fixed bug with using `BIND_ENUM_CONSTANT` instead of `BIND_BITFIELD_FLAG` (now the error can be found by `make_rst.py`).

https://github.com/godotengine/godot/blob/28cca66d2cfe25d6d7bf5a2a26ab7bd366029669/scene/main/node.h#L709

https://github.com/godotengine/godot/blob/28cca66d2cfe25d6d7bf5a2a26ab7bd366029669/scene/main/node.cpp#L49-L52

Is there any reason why some `VARIANT_*_CAST` are in `*.cpp` while most are in `*.h`?

`VARIANT_(ENUM|BITFIELD)_CAST` in `*.cpp` - 11 matches

```
core/io/ip.cpp: 1
 38:  1: VARIANT_ENUM_CAST(IP::ResolverStatus);
core/io/xml_parser.cpp: 1
 37:  1: VARIANT_ENUM_CAST(XMLParser::NodeType);
core/os/time.cpp: 2
 55:  1: VARIANT_ENUM_CAST(Month);
 56:  1: VARIANT_ENUM_CAST(Weekday);
scene/2d/line_2d.cpp: 3
 38:  1: VARIANT_ENUM_CAST(Line2D::LineJointMode)
 39:  1: VARIANT_ENUM_CAST(Line2D::LineCapMode)
 40:  1: VARIANT_ENUM_CAST(Line2D::LineTextureMode)
scene/main/node.cpp: 4
 49:  1: VARIANT_ENUM_CAST(Node::ProcessMode);
 50:  1: VARIANT_ENUM_CAST(Node::ProcessThreadGroup);
 51:  1: VARIANT_BITFIELD_CAST(Node::ProcessThreadMessages);
 52:  1: VARIANT_ENUM_CAST(Node::InternalMode);
```

`VARIANT_(ENUM|BITFIELD)_CAST` in `*.h` - 625 matches

akien-mga reviewed (approved):
Looks good to me. Needs a rebase.

dalexeev committed:
Display `BitField[Enum]` in docs to distinguish from `Enum`

akien-mga commented:
Thanks!
